### ISSUE TITLE
Unable to edit scripts through the visual code extension

### ISSUE THREAD

**Issue by HansenRene** [ironmansoftware/powershell-universal#1174]
### Steps to Reproduce

I can open a script, edit it and save it without errors. but the change is not reflected in the admin console, nor does git react to it.

Editing Dashboards works just fine through the extension.
Thought the issue might relate to the script being placed in a folder, but alas.

Don't know if its related, but my scripts folder in /admin/settings/data is pointed to another location than blank (UniversalAutomation\Repository\Scripts).

### Expected behavior

```console
it working would be nice. 
the extension throwing an error would be even better actually. I've been editing that script for 3 hours :D
```

### Actual behavior

```console
N/A
```

### Environment data

PSU 2.11.0
Extension 2.10.2

**Comment by adamdriscoll** [comment]
I have been unable to reproduce the issue with saving. I certainly can cause an issue (like stopping PSU) that will cause the file to "save" without saving. I added some error handling so you will now get a popup if that happens. 

I'll publish a fixed version and you can give it another shot and see if it gives a good error message so I can resolve the issue with it not saving.

**Comment by HansenRene** [comment]
No luck, updated both the extension and VScode. 

The file is saved to (C:\Users\USER\AppData\Local\Temp\.universal.code.script\Test\Test2.ps1), but no errors appear, and the script is not updated in the admin dashboard within any period of wait.
reopening the script just overwrites it.

My settings looks like this:
`"powerShellUniversal.connections": [

    {
        "name": "Beta.MiddlewareDashboard",
        "samplesDirectory": "C:\\Users\\userHERE\\AppData\\Roaming\\PSU-MWb-Samples",
        "url": "https://urlHERE.org",
        "appToken": "tokenHERE"
    },
    {
        "name": "Beta.SupportDashboard",
        "samplesDirectory": "C:\\Users\\userHERE\\AppData\\Roaming\\PSU-SDb-Samples",
        "url": "https://urlHERE.org",
        "appToken": "tokenHERE"
    }
],
"powerShellUniversal.checkModules": false,`

**Comment by adamdriscoll** [comment]
It seems like it's not trying to save the file. Just as an explanation of what's happening in the extension:

We are watching for documents with the specified path to be saved. 

https://github.com/ironmansoftware/universal-code/blob/master/src/commands/scripts.ts#L42

Once they are saved, we get the latest version of the script. 

https://github.com/ironmansoftware/universal-code/blob/master/src/commands/scripts.ts#L47

Update the content and then save the script. 

https://github.com/ironmansoftware/universal-code/blob/master/src/commands/scripts.ts#L49

In both cases, we're just calling the PSU API. 

https://github.com/ironmansoftware/universal-code/blob/54729efd2d9ad0950173d1d28ea267f9f15517f2/src/universal.ts#L294

https://github.com/ironmansoftware/universal-code/blob/54729efd2d9ad0950173d1d28ea267f9f15517f2/src/universal.ts#L302

One thing I would recommend is verifying that you can call things through the REST API using the same app token and URL as VS Code. 

```
$AppToken = "tokenHERE"
$Script = Invoke-RestMethod http://urlHERE.org/api/v1/script/path/Test/Test.ps1 -Headers @{ Authorization = "Bearer $AppToken" } 
$Script.Content = "Hello from PS!"
Invoke-RestMethod http://urlHERE.org/api/v1/script/ -METHOD put -Body ($Script | ConvertTo-Json)  -Headers @{ Authorization = "Bearer $AppToken" } -ContentType "application/json" 
```

I bet this works fine. I have a feeling what is actually happening is the file isn't being detected as changed for some reason and we aren't making the call at all. 

If you want to see what web requests are coming into your PSU server, you can set the Microsoft log level to Debug and it will log out every web request coming into PSU. Obviously, with a very busy server it's a pretty noisy log. 

![image](https://user-images.githubusercontent.com/1256531/169834534-8ce83fc8-a9a5-4b8a-a802-4c84c0525115.png)

**Comment by HansenRene** [comment]
running line 1-3 works fine, but on line 4
`Invoke-RestMethod https://beta.myURLHERE.org/api/v1/script/ -METHOD put -Body ($Script | ConvertTo-Json)  -Headers @{ Authorization = "Bearer $AppToken" } -ContentType "application/json" `

I get a buuuunch of stuff back that includes Twitters MIT license and stuff? 😅

**Comment by adamdriscoll** [comment]
Whoa. Weird. 

Is it returning HTML? A 401, 403 or 404 redirect couple potentially return an HTML page depending on your authentication. This might explain why you aren't seeing errors. If it's redirecting to somewhere else instead of returning an error code, then it's going to return a 200 success code but redirect over to something other than the API we're attempting to call. 

Can you grab a debug log and send it to me at adam@ironmansoftware.com? I wanna see what the heck this thing is doing.

**Comment by HansenRene** [comment]
Came across this issue on another API. (on another PSU instance)
A Dashboard wouldn't load data when an api that was used had auth enabled. Token had Admin role.
I tried it manually and I get the same monstrous output.

**Comment by HansenRene** [comment]
In this case though, it happens because of a malformed URI in the REST call. maybe that can be relevant in your troubleshooting. 

This, with empty variables, gives me the HTML output.
`Invoke-RestMethod https://beta.ourdashboard.domain.org/$GetADDataAPI/$Query/$Session:LookupDomain/ -Headers @{ Authorization = "Bearer TOKENHERE" }`

This Gives me an 404 as expected
`Invoke-RestMethod https://beta.ourdashboard.domain.org/Get-ADDataRENHAN/$Query/$LookupDomain/  -Headers @{ Authorization = "Bearer TOKENHERE" } `

These return HTML
`Invoke-RestMethod https://beta.ourdashboard.domain.org`
`Invoke-RestMethod https://beta.ourdashboard.domain.org -Headers @{ Authorization = "Bearer TOKENHERE" }`

Is there variables in the script save function that doesnt get populated maybe?

**Comment by adamdriscoll** [comment]
The save script call is static and we don't put variables in the URL. 

![image](https://user-images.githubusercontent.com/1256531/172716061-dd32e1e9-a485-4d4c-af58-52ea0a05fc1e.png)

The fact that you can reproduce this from PowerShell means that we have something jacked up on the server and I am struggling to figure out what that might be. 

Just a heads up on how I'm testing this. 

Running PSU just as a process and listening over HTTP and HTTPS. 

![image](https://user-images.githubusercontent.com/1256531/172717352-c315f07b-942b-4d2d-b566-c42194729b83.png)

I have a nested script base folder. 

![image](https://user-images.githubusercontent.com/1256531/172716781-659c05a7-45da-491b-92d8-8f242014ab5f.png)

I've created scripts at a couple different levels. 

![image](https://user-images.githubusercontent.com/1256531/172716871-c974f1c7-c808-4f8c-917a-1ab150a6ad78.png)

I'm using OpenID Connect for auth. 

![image](https://user-images.githubusercontent.com/1256531/172716929-5e0b7cb2-279f-4861-8da2-2c6f2c5302c8.png)

I've generated an app token for my user. 

![image](https://user-images.githubusercontent.com/1256531/172716970-8be44526-6154-4884-b54a-f95108347294.png)

I'm executing via the REST API directly with no errors. 

````
PS C:\Users\adamr> $Script = Invoke-RestMethod https://localhost:5001/api/v1/script/Folder/Test/Script2.ps1 -Headers @{ Authorization = "Bearer $AppToken" }
PS C:\Users\adamr> $Script.Content = "Hello from PS!"
PS C:\Users\adamr> Invoke-RestMethod https://localhost:5001/api/v1/script/ -METHOD put -Body ($Script | ConvertTo-Json)  -Headers @{ Authorization = "Bearer $AppToken" } -ContentType "application/json"
````

And from VS code. 

![image](https://user-images.githubusercontent.com/1256531/172717159-e856b531-7d72-45e1-9fbf-93740793d438.png)

I'm using the latest 2.12.0 build and latest version of the VS Code extension. Spot any differences that I should investigate?

**Comment by HansenRene** [comment]
I finally got it.

So in an earlier version of PSU (I guess, don't know that's fixed) there was an issue if you visited URLs that ended with a slash, it would add a second slash and of course that didn't work.

So I made a rewrite rule to remove trailing slashes.

  `<rule name="Remove trailing slash" enabled="false" stopProcessing="true">
    <match url="(.*)/$" />
    <conditions logicalGrouping="MatchAll" trackAllCaptures="false">
        <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="true" />
        <add input="{REQUEST_FILENAME}" matchType="IsDirectory" negate="true" />
    </conditions>
    <action type="Redirect" url="{R:1}" redirectType="Permanent" />
  </rule>`

And that worked for solving that.

But it does mess with other stuff that I didn't use. Besides the issue with script saving, it also kills any attempt to invoke API's where you add a Slash at the end of the URI - but only on API's with authentication.

**Comment by adamdriscoll** [comment]
So just to understand correctly. Are you still having the issue with the slashes or does that seem to be resolved? If not, we should fork this issue and get that fixed.

**Comment by HansenRene** [comment]
In the entire 2 minutes I had to test during lunch, I found no issue. I'll make another issue of it if I find anything.


Regarding this issue, while it was clearly caused by url rewrite, some kind of confirmation of successful save / error would be real nice.

Hent Outlook til Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by adamdriscoll** [comment]
I agree. I'll see what we can do. The problem at the moment is that it's returning a 200 from the server but we should be parsing the result to validate it's actually hitting the correct URL. 

UPDATE: This has been resolved. I had to update the PSU server since it wasn't actually returning anything except 200 OK. Now it will return the updated script after saving it. The PSU extension will check that return value to ensure that the content of the script matches the local content of the script.
